src/masternodes/rpc_icxorderbook.cpp

```diff
@@ -0,0 +1,1479 @@
+
+            if (pcustomcsview->HasICXSubmitDFCHTLCOpen(submitdfchtlc.offerTx))
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "dfc htlc already submitted!");
+            if (pcustomcsview->HasICXSubmitEXTHTLCOpen(submitdfchtlc.offerTx))
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("offer (%s) needs to have dfc htlc submitted first, but external htlc already submitted!",
+                        submitdfchtlc.offerTx.GetHex()));
+
+            if (submitdfchtlc.timeout < CICXSubmitDFCHTLC::MINIMUM_TIMEOUT)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid parameters, argument \"timeout\" must be greater than %d", CICXSubmitDFCHTLC::MINIMUM_TIMEOUT - 1));
+
+        }
+        else if (order->orderType == CICXOrder::TYPE_EXTERNAL)
+        {
+            authScript = offer->ownerAddress;
+
+            CTokenAmount balance = pcustomcsview->GetBalance(offer->ownerAddress,order->idToken);
+            if (balance.nValue < offer->amount)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Not enough balance for Token %s on address %s!",
+                        pcustomcsview->GetToken(order->idToken)->CreateSymbolKey(order->idToken), ScriptToString(offer->ownerAddress)));
+
+            if (pcustomcsview->HasICXSubmitDFCHTLCOpen(submitdfchtlc.offerTx))
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "dfc htlc already submitted!");
+            auto exthtlc = pcustomcsview->HasICXSubmitEXTHTLCOpen(submitdfchtlc.offerTx);
+            if (!exthtlc)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("offer (%s) needs to have ext htlc submitted first, but no external htlc found!",
+                            submitdfchtlc.offerTx.GetHex()));
+
+            CAmount calcedAmount(static_cast<CAmount>((arith_uint256(exthtlc->amount) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
+            if (submitdfchtlc.amount != calcedAmount)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("cannot make dfc htlc with that amount, amount must be equal to calculated exthtlc amount - %s != %s",
+                                ValueFromAmount(submitdfchtlc.amount).getValStr(), ValueFromAmount(calcedAmount).getValStr()));
+
+            if (submitdfchtlc.hash != exthtlc->hash)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid hash, dfc htlc hash is different than extarnal htlc hash - %s != %s",
+                        submitdfchtlc.hash.GetHex(),exthtlc->hash.GetHex()));
+            if (submitdfchtlc.timeout < CICXSubmitDFCHTLC::MINIMUM_2ND_TIMEOUT)
+                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid parameters, argument \"timeout\" must be greater than %d", CICXSubmitEXTHTLC::MINIMUM_2ND_TIMEOUT - 1));
+            if (submitdfchtlc.timeout >= (exthtlc->creationHeight + (exthtlc->timeout * 100 / 6 )) - targetHeight)
```

Comment:
I'm curious why we are using `100 / 6` here, DFC block is every 30 seconds, and bitcoin block is every 10 minutes, shouldn't it be a 20 <-> 1 conversion ratio?

Note that we are altering the difficulty adjustment rate in this release too, so we are expecting the DFC blocktime to stabilise at 30 seconds instead of the current 37 seconds.

Reply:
I was using this as per notes we had from two weeks ago. I also checked with U-Zyn should I use 100 blocks per hour for DFI.
I can change this just let me know which is the number we are targeting. This is only theoretically as we cannot know how much blocks exactly will be in an hour.